### PR TITLE
Implement Time macro

### LINKED ISSUE #160: Time macro similar to CUDA.@time
Is a time macro that works similarly to CUDA.@time a feature that would be welcome? 

I'd like to file an issue related to memory management (https://github.com/odsl-team/julia-ml-from-scratch/issues/2) and having that macro would give me a better idea of what's going on.

I would be happy to implement it but I would need a bit of guidance. 
Or is there a better existing way of doing this? Profiling seems overkill.

### PR DESCRIPTION
**Pull request by christiangnrd**
@maleadt This is my first go at it. There was already a memory.jl file so I just created pool.jl and put the code there (like it is in CUDA.jl). Would you rather I move it into memory.jl?

I wasn't quite sure what you meant by low-level MTLBuffer constructor so I left them all in where they were.

I copied over the tests from CUDA.jl except for one which didn't work because of an unsafe_wrap function not found but it also does not seem like a relevant test given the unified memory model.

Close #160

### CODE REVIEW

File: src/pool.jl [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+Note that `Private` buffers can't be directly accessed from the CPU, therefore you cannot
+use this option if you pass a ptr to initialize the memory.
+"""
+function alloc(args...; kwargs...)
+    time = Base.@elapsed begin
+        buf = MTLBuffer(args...; kwargs...)
+    end
+
+    alloc_stats.alloc_count += 1
+    alloc_stats.alloc_bytes += args[2]
```
Comment (christiangnrd) [inline src/pool.jl:55]: I don't love the way I did this. Is there a better way?

Comment (maleadt) [reply src/pool.jl:55]: I would make the function more concretely typed. We only ever invoke it with `dev, bytesize` as the first two args.

**Comment by maleadt** [comment]
Nice, thanks! Maybe add `@allocated` too?

Also, when looking over your changes, I figured we should probably use atomics for the stats. So I changed that in https://github.com/JuliaGPU/CUDA.jl/pull/1884; you could mirror that here if you want.

File: src/pool.jl (hunk 1 above)
Comment (christiangnrd) [reply src/pool.jl:55]: Done

**Comment by christiangnrd** [comment]
`@allocated` was already in!

I've mirrored the atomics changes.

File: src/pool.jl [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+        val
+    end
+end
+
+macro timed(ex)
+    quote
+        while false; end # compiler heuristic: compile this block (alter this if the heuristic changes)
+
+        # @time(d) might surround an application, so be sure to initialize CUDA before that
```
Comment (max-Hawkins) [inline src/pool.jl:174]: Should the CUDA-specific comments be removed?

Comment (christiangnrd) [reply src/pool.jl:174]: yes

**Comment by maleadt** [comment]
https://github.com/JuliaGPU/CUDA.jl/pull/1885

Also, because Metal.jl only supports Julia 1.8+, we should be able to use `@atomic` instead. That should make the code much cleaner.

**Comment by christiangnrd** [comment]
I've made the change to @atomic and the tests pass. However, I'm not sure if I've implemented it correctly. Please let me know if anything needs to be changed.

File: src/pool.jl [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    @atomic free_bytes::Int
+
+    @atomic total_time::Float64
+    function AllocStats()
+        new(Int(0), Int(0),
+            Int(0), Int(0),
+            Float64(0.0))
+    end
+
+    function AllocStats(alloc_count::Integer, alloc_bytes::Integer,
```
Comment (maleadt) [inline src/pool.jl:21]: You don't need this this constructor now that we don't use `Threads.Atomic{...}`. Just move out the parameter-less one and rely on the default constructor, which already does conversions.

**Pull request merged**